cpp/src/stream_compaction/drop_duplicates.cu

```diff
@@ -0,0 +1,303 @@
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy_if.cuh>
+#include <cudf/stream_compaction.hpp>
+#include <cudf/detail/stream_compaction.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/detail/gather.hpp>
+#include <thrust/logical.h>
+#include <thrust/functional.h>
+#include <thrust/execution_policy.h>
+#include <algorithm>
+#include <cmath>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+/*
+ * unique_copy copies elements from the range [first, last) to a range beginning
+ * with output, except that in a consecutive group of duplicate elements only
+ * depending on last argument keep, only the first one is copied, or the last
+ * one is copied or neither is copied. The return value is the end of the range
+ * to which the elements are copied.
+ */
+template<typename Exec,
+         typename InputIterator,
+         typename OutputIterator,
+         typename BinaryPredicate,
+    typename IndexType = typename
+  thrust::iterator_difference<InputIterator>::type>
+  OutputIterator unique_copy(Exec&& exec,
+                             InputIterator first,
+                             InputIterator last,
+                             OutputIterator output,
+                             BinaryPredicate comp,
+                             const duplicate_keep_option keep)
+{
+  IndexType n = (last-first)-1;
+  if (keep == duplicate_keep_option::KEEP_FIRST) {
+      return thrust::copy_if(exec,
+              first,
+              last,
+              thrust::counting_iterator<IndexType>(0),
+              output, 
+              [first, comp, n] __device__ (IndexType i) mutable {
+              return (i == 0 || !comp(first[i], first[i-1]));
+              }); 
+  } else if (keep == duplicate_keep_option::KEEP_LAST) {
+      return thrust::copy_if(exec,
+              first,
+              last,
+              thrust::counting_iterator<IndexType>(0),
+              output, 
+              [first, comp, n] __device__ (IndexType i) mutable {
+              return (i == n || !comp(first[i], first[i+1]));
+              });
+  } else {
+      return thrust::copy_if(exec,
+              first,
+              last,
+              thrust::counting_iterator<IndexType>(0),
+              output, 
+              [first, comp, n] __device__ (IndexType i) mutable {
+              return (i == 0 || !comp(first[i], first[i-1])) 
+                  && (i == n || !comp(first[i], first[i+1]));
+              });
+  }
+}
+
+auto get_unique_ordered_indices(cudf::table_view const& keys,
+                                cudf::column &unique_indices,
+                                duplicate_keep_option const& keep,
+                                bool const& nulls_are_equal = true,
+                                rmm::mr::device_memory_resource* mr =
```

Comment:
Likewise, this API only allocates temporary memory. Therefore, it should not accept a MR.